### ISSUE TITLE
perf(torch): reduce CPML traffic with axis-sparse residuals

### ISSUE THREAD

**Issue by ruddyscent** [ruddyscent/gmes#150]
## Parent and motivation

Child of #145. The retained Torch CPU graph improvements still leave CPML as the dominant GMES-controlled allocation and memory-traffic source in `cpu-crossover-2d`.

The current compact CPML path processes 10,379 component targets in six buckets. Inductor functionalization materializes two full `N x 2` state outputs plus one `N` field-target output, exactly `415,160 B/step`, or 80.3% of the accepted short trace's allocated bytes. The dense stencil already traverses CPML cells but currently multiplies their update by zero, after which the compact path gathers both curl directions again.

## Goal

Reduce CPU-compiled CPML memory traffic by applying the ordinary curl through the dense field-aligned coefficient plane and lowering only active CPML axes as sparse residual updates. Preserve the native CPML equations, public field behavior, logical PML state snapshots, and versioned checkpoint interface.

Keep eager, CUDA, distributed, and UPML execution on their existing representations unless separate evidence supports changing them.

## Proposed representation

For each CPML component target, evaluate the ordinary dense update using the immediate underlying material's high-frequency inverse. For each active curl axis, apply only:

```text
psi = b * psi + c * derivative
field += dt * direction * inv_base * axis_sign
         * ((1 / kappa - 1) * derivative + psi)
```

An axis is structurally active when `c != 0` or `kappa != 1`. Store active-axis state in contiguous one-dimensional buffers, while reconstructing the canonical `(target_count, 2)` logical state at snapshot and checkpoint boundaries. Reject checkpoints that inject nonzero values into omitted structurally inactive lanes rather than silently discarding them.

## Implementation work

- [x] Add planner descriptors for active CPML axes and fill the dense inverse at CPML targets.
- [x] Add CPU-compiled sparse residual execution with contiguous per-axis state and fixed scratch.
- [x] Preserve canonical PML snapshot/checkpoint keys, shapes, target order, and fixed-address restoration.
- [x] Cover custom `kappa_max`, mixed underlying media, collapsed dimensions, paired-real Bloch fields, float32, and long-run absorption.
- [x] Measure generated code, allocations, physical state bytes, and 1-/4-thread end-to-end timing against commit `1c6b2fb`.
- [x] Run the relevant full regression suite and build, then retain the representation only if the measured CPU result materially improves without numerical or memory regressions.

## Acceptance criteria

- Fields and logical CPML state remain within the frozen native-oracle tolerances at the existing 1/2/5/20/100-step capture points, including custom kappa and mixed-underlying cases.
- `pml_state_snapshot()` and simulation checkpoint version 1 retain their canonical logical `(N, 2)` representation; restore does not replace live buffers.
- Local compiled CPU execution retains exactly two compiled regions with no warm graph breaks, recompilations, host transfers, or storage-address changes.
- Physical CPML state and recurring CPML allocation bytes are reduced by at least 50% on `cpu-crossover-2d`.
- The selected representation improves, or at minimum does not regress outside measurement noise, the fixed 1-thread and physical-core timing samples.
- Eager, UPML, CUDA, and distributed selection and semantics remain unchanged; unavailable accelerator validation is reported explicitly.
- Raw benchmark/profiler evidence and generated-code findings are attached before disposition.

## Preliminary local evidence

An uncommitted prototype in an isolated worktree produced the exact accepted final checksum and clean compiler counters under a short 1-thread diagnostic:

| Metric | Current `1c6b2fb` | Prototype |
|---|---:|---:|
| Median ms/step | 1.4500 | 1.2845 |
| Positive allocation events | 232 | 212 |
| Allocated bytes / 2 steps | 1,034,292 | 296,132 |
| Maximum allocation | 30,144 | 6,400 |
| Physical CPML state | 166,064 B | 46,080 B |

These numbers are diagnostic only and must be reproduced after tests and review; they do not alter #145's unchanged native timing gate.

**Comment by ruddyscent** [comment]
## Insight: CPML traffic is sparse by curl axis, not only by cell

The implemented decomposition preserves the native equation while moving the ordinary curl into the already-running dense stencil:

```text
dense:    inv_base * d
residual: inv_base * ((1 / kappa - 1) * d + psi)
sum:      inv_base * (d / kappa + psi)
```

The sign is applied independently for the two curl terms, including the reversed magnetic ordering. This makes the residual state sparse by **axis**. In `cpu-crossover-2d`, 10,379 logical CPML cells expose 20,758 logical state lanes, but only 5,760 lanes are structurally active (`c != 0 || kappa != 1`). The hot physical state therefore falls from 166,064 B to 46,080 B (-72.25%).

Generated-code inspection of the preliminary two-step trace attributes the complete allocation reduction to CPML:

- CPML recurring output bytes: 415,160 B/step -> 46,080 B/step (-88.90%).
- CPML allocation events: 18/step -> 8/step (-55.56%).
- Generated CPML loops: 22 -> 15; scalar loop iterations: 38,056 -> 10,880 (-71.4%).
- Non-CPML allocation bins are unchanged.

The remaining eight CPML allocations are one field-sized residual result per active axis. A prior `index_add_` ablation removed more of that traffic but regressed end-to-end CPU time, so the current gather/add/scatter residual is retained pending the clean official measurement.

The physical layout is intentionally hidden at persistence boundaries. `pml_state_snapshot()` and simulation checkpoint version 1 reconstruct the canonical `(target_count, 2)` state in original target order, restore into fixed buffers, and reject nonzero values in omitted inactive lanes. The documented checkpoint APIs are the compatibility contract; internal sparse buffers remain non-persistent in ordinary `nn.Module.state_dict()`. Planner `estimated_bytes` was also corrected to count the full NumPy arrays that remain resident for diagnostics, separately from the smaller finalized Torch plan.

These figures are diagnostic evidence from the dirty prototype, not final disposition evidence. The clean-commit 1-/4-thread contract, profiler hashes, full regression suite, and build will be attached separately.

**Comment by ruddyscent** [comment]
## Final acceptance evidence

Candidate: `92be7a12cb806fbb69ffe0d07f9d3102caa49937` (clean worktree), compared with parent commit `1c6b2fb320f67f2c7433e7ec5e7a7a09726a5b04`. Both CPU modes used the frozen `torch-cpu-acceptance-v3` contract: 5 warmup steps, 15 independently restored repetitions of 100 steps, and a 5-step profiler trace. The pinned native summary SHA-256 was `a81b87e8cb2870d2fbbf1bbb7409c98d10780da30eb23cb5965fce39b9109fb6`.

| Metric | `1c6b2fb` | `92be7a1` | Change |
|---|---:|---:|---:|
| 1-thread median | 1.456151 ms/step | 1.222634 ms/step | -16.04% |
| 1-thread Torch/native | 1.7154x | 1.4403x | improved |
| 4-thread median | 1.307702 ms/step | 1.172482 ms/step | -10.34% |
| 4-thread Torch/native | 1.9642x | 1.7611x | improved |
| Physical CPML state | 166,064 B | 46,080 B | -72.25% |
| CPML recurring allocations | 415,160 B/step | 46,080 B/step | -88.90% |
| Generated CPML tensor traffic | 3,819,472 B/step | 829,440 B/step | -78.28% |

Raw candidate samples in ms/step:

```text
1T: 1.222634, 1.228907, 1.222229, 1.222684, 1.248606,
    1.226346, 1.206942, 1.247492, 1.223681, 1.202131,
    1.215043, 1.197849, 1.207321, 1.196621, 1.254129
4T: 1.172482, 1.201565, 1.236904, 1.144916, 1.176637,
    1.173127, 1.164435, 1.139408, 1.153566, 1.151321,
    1.149972, 1.187568, 1.175458, 1.162426, 1.300983
```

### Generated code and profiler findings

- Both thread modes retained exactly two compiled regions, zero warm graph breaks/recompilations/cache misses, zero host/device transfers, the expected source-only external indexed writes, and stable storage addresses. The final checksum was `267.52060706901773` in both modes.
- Generated CPML code fell from 22 to 15 C++ loops and from 38,056 to 10,880 row-pass extents per step. The 4-thread wrapper keeps these small residual loops under `omp single`, so the traffic reduction—not CPML loop parallelism—explains the improvement.
- CPML allocation sites fell from 18 to 8 per step. The new sites are four 800-element and four 640-element float64 temporaries, totaling 46,080 B/step.
- Across each 5-step official trace, total positive allocations fell from 2,585,730 B/580 events to 740,330 B/530 events. All non-CPML histogram bins were identical; the complete 1,845,400 B reduction was CPML-attributable. Trace hashes: 1T `7c96d9360d98fa6aec33c70f809cf8d95d578d541fe5d6ba77c8bd20f8ce0d10`; 4T `5ae135e5e642870d52499d5a6842b07afb68faa67feed8171464af681c708d89`.
- Physical CPML state contains 5,760 active axis lanes rather than 20,758 logical lanes. Snapshot and checkpoint boundaries still expose the canonical target-order `(N, 2)` state.

### Correctness and compatibility

- Full test suite: **398 passed**, with **7 CUDA-only skips** because this host's PyTorch build reports no CUDA devices.
- Native-oracle CPML comparisons passed at 1/2/5/20/100 steps for float64 and float32, custom `kappa_max`, heterogeneous underlying media, collapsed dimensions, and paired-real Bloch fields; the compiled long-run absorption regression also passed.
- Standard `state_dict()` and versioned simulation checkpoint v1 round trips preserve canonical logical CPML state and live buffer addresses. Nonzero omitted-lane state is rejected.
- Extreme float32 `kappa_max=1e8` remains on the existing compact full-curl representation when base-plus-residual reconstruction would lose the direct reciprocal coefficient.
- Eager, UPML, CUDA, and distributed selection paths were not changed. CUDA execution itself was unavailable and is reported rather than inferred.
- `uv build` succeeded for the sdist and CPython 3.14 Linux wheel; only the existing SWIG string-truncation warnings were emitted.

This satisfies #150's scoped acceptance criteria: physical state and recurring CPML allocation reductions exceed 50%, both fixed CPU timing modes materially improve, numerical/persistence behavior is preserved, and raw evidence is recorded. It does **not** satisfy parent #145's separate 1.05x native-parity or zero-recurring-allocation goals: the remaining Torch/native ratios are 1.4403x and 1.7611x, and each diagnostic case still reports `recurring_allocations_zero=false`.

**Comment by ruddyscent** [comment]
## Integration gate insight

PR #152 targets the unfinished parent branch `perf/145-reshape-inductor-cpu-graphs`, not `master`, so the repository's CI workflow intentionally does not create the required Ubuntu/macOS PR checks. This is confirmed by `.github/workflows/ci.yml` and its regression test; CI has no `workflow_dispatch` entry, so those two check runs cannot be attached to this non-master PR without changing its base or workflow contract.

The checks that can be run safely are complete on final commit `718e6777ac7dbe6d6b1227c51fd0e26f61ccc79e`:

- manually dispatched CodeQL run [33097055687](https://github.com/ruddyscent/gmes/actions/runs/33097055687) passed both Python and C/C++ analysis, with no open branch alerts;
- the complete local suite passed 398 tests with 7 CUDA-only skips;
- the sdist and CPython 3.14 Linux wheel built successfully;
- all seven Python files owned by #150 pass pinned Black 26.5.1 and isort 8.0.1 checks, and pylint reports 10.00/10;
- PR #152 is cleanly mergeable and has no review conversations.

Repository-wide formatting still reports pre-existing #145-branch debt in six files untouched by #150 (`gmes/torch_dm2.py`, three benchmark modules, and two tests). Those cosmetic fixes were deliberately not mixed into this performance child issue.

The squash merge was not performed because the repository workflow requires reported Ubuntu/macOS checks and the non-master PR cannot produce them. The remaining integration decision is whether to explicitly allow merging #152 into its parent development branch with the local/full-suite, build, and CodeQL evidence above, leaving the master-targeted parent PR to run the required cross-platform CI later.

**Comment by ruddyscent** [comment]
Implementation is now squash-merged by #152 into the parent #145 development branch as commit `0719198` (verified GitHub signature).

The child acceptance criteria are satisfied by the evidence already attached above. This issue remains open only because GitHub does not apply `Closes #150` when a pull request merges into a non-default parent development branch, and I am not manually closing it without separate authorization. When #145 reaches `master`, revisit this issue's disposition; the implementation itself no longer blocks the parent branch.

**Comment by ruddyscent** [comment]
## Final disposition

PR #152 was squash-merged into the #145 development branch as `071919869c0c2ff20b0bdf77c4d5e8b111f085a0`.

The scoped acceptance criteria are complete. The axis-sparse CPML residual reduced physical CPML state by 72.25%, recurring CPML allocation by 88.90%, and generated CPML tensor traffic by 78.28%, while improving the fixed 1T and 4T timings by 12.58% and 9.74%. Numerical, logical persistence, two-region, storage-identity, build, test, style, and CodeQL evidence is preserved above.

The remaining parent timing and zero-recurring-allocation failures belong to #145 and do not leave work outstanding in this child. Closing as completed; final master integration will occur through the accumulated #145 pull request.
